walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,87 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  /**
+   * @param events a collection of events going on during the day
+   * @param request user created request for an event that includes Time and attendees
+   * @return a list of available times to schedule a meeting
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> empty = new ArrayList<>();
+    ArrayList<String> attendees = new ArrayList<String>(request.getAttendees()); // list for mandatory attendees
+    ArrayList<String> allAttendees = new ArrayList<String>(attendees);
+    allAttendees.addAll(request.getOptionalAttendees());
+    // return whole day if there are no attendees
+    if (allAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+    //return empty If there are no events during day(gaps), but there are optional attendees
+    if(events.isEmpty() && !request.getOptionalAttendees().isEmpty()){
+      return empty;
+    }
+    // When there are no available times with optional attendees, set available times for mandatory employees only
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>(
+        availableTimesForAttendees(events, allAttendees, request.getDuration()));
+    if(!availableTimes.isEmpty()){
+      return availableTimes;
+    }
+    else{
+      availableTimes = new ArrayList<TimeRange>(availableTimesForAttendees(events, attendees, request.getDuration()));
+    }
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if(!attendees.isEmpty()){
```

Comment:
can't we check for attendees empty before running line "availableTimes = new ArrayList<TimeRange>(availableTimesForAttendees(events, attendees, request.getDuration()));"

that will prevents unnecessary calls to availableTimesForAttendees